File: backend/app/services/dashboard_service.py
from sqlalchemy.orm import Session

from app.repositories.dashboard_repository import (
    DashboardRepository,
)


class DashboardService:

    # ---------------------------------------------------------
    # GET DASHBOARD
    # ---------------------------------------------------------

    @staticmethod
    def get_dashboard(
        db: Session,
    ):

        data = DashboardRepository.get_dashboard(
            db,
        )

        latest_notifications = []

        for notification in data["latest_notifications"]:

            latest_notifications.append(
                {
                    "id": notification.id,
                    "exam_name": notification.exam_name,
                    "notification_no": notification.notification_no,
                    "organization": notification.organization,
                    "application_end": (
                        str(notification.application_end)
                        if notification.application_end
                        else None
                    ),
                }
            )

        return {

            "total_notifications":
                data["total_notifications"],

            "open_notifications":
                data["open_notifications"],

            "closed_notifications":
                data["closed_notifications"],

            "organizations":
                data["organizations"],

            "latest_notifications":
                latest_notifications,

            "category_distribution":
                data["category_distribution"],

            "organization_distribution":
                data["organization_distribution"],
        }

File: frontend/src/services/dashboardService.ts
import api from "./api";

export interface DashboardResponse {
  total_notifications: number;
  open_notifications: number;
  closed_notifications: number;
  organizations: number;

  latest_notifications: {
    id: number;
    exam_name: string;
    organization: string;
    notification_no: string;
    application_end: string;
  }[];
}

export async function getDashboard() {
  const { data } = await api.get<DashboardResponse>(
    "/api/v1/dashboard"
  );

  return data;
}